user_group_popover: Convert module to typescript.

sahil839 opened the pull request:
<!-- Describe your pull request here.-->
This is needed for #24580.
<!-- Issue link, or clear description.-->

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [ ] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

sahil839 commented on web/src/user_group_popover.ts:
```diff
@@ -99,13 +107,14 @@ export function toggle_user_group_info_popover(element, message_id) {
         const $row = $elt.closest(".message_row");
-        const message = message_lists.current.get(rows.id($row));
+        assert(message_lists.current);
+        const message = message_lists.current.get(rows.id($row)!)!;
```
I assumed here that `rows.id` and `message_lists.current.get` will always be defined and non-null, since this handler is called when a user clicks on a mention pill from a message.

andersk replied:
This is not a performance-sensitive hot loop, and the reasoning required to establish that these values aren’t undefined or null is complicated and very nonlocal, so for either or both of those reasons, don’t use a non-null assertion (`!`).

andersk commented on web/src/user_group_popover.ts:
```diff
@@ -99,13 +107,14 @@ export function toggle_user_group_info_popover(element, message_id) {
         const $elt = $(e.currentTarget);
         const $row = $elt.closest(".message_row");
-        const message = message_lists.current.get(rows.id($row));
+        assert(message_lists.current);
```
Never write ambiguous truthiness checks that make it unclear which falsy value you’re guarding against.

```suggestion
        assert(message_lists.current !== undefined);
```

andersk commented on web/src/user_group_popover.ts:
```diff
@@ -8,32 +10,35 @@ import * as hash_util from "./hash_util";
+export function hide(): void {
     if (is_open()) {
-        user_group_popover_instance.destroy();
+        user_group_popover_instance!.destroy();
```
Don’t use an unchecked non-null assertion (`!`); fix the code so TypeScript understands why the value is non-null.

```diff
-    if (is_open()) {
+    if (user_group_popover_instance !== undefined) {
```

andersk commented on web/src/user_group_popover.ts:
```diff
@@ -8,32 +10,35 @@ import * as hash_util from "./hash_util";
     }
 
-    const $popover = $(user_group_popover_instance.popper);
+    const $popover = $(user_group_popover_instance!.popper);
```
Same.

andersk commented on web/src/user_group_popover.ts:
```diff
@@ -42,20 +47,23 @@ function get_user_group_popover_items() {
     }
     const $elt = $(element);
-    const user_group_id = Number.parseInt($elt.attr("data-user-group-id"), 10);
+    const user_group_id = Number.parseInt($elt.attr("data-user-group-id")!, 10);
```
Same.

sahil839 commented:
Updated the PR. I have replaced the use of `!` operator with `assert` statements considering that we don't expect the values to be `undefined` or `null`. Do we instead want to have an `if` block with early return or something like that?

timabbott commented:
Assert is usually better than if/else clauses if we think the condition should only fail due to a bug in the software. Having an if/else clause suggests that we think that is a possibility, is messy for enforcing test coverage (since it's hard to test impossible scenarios), etc.

timabbott commented on web/src/user_group_popover.ts:
```diff
@@ -42,21 +47,26 @@ function get_user_group_popover_items() {
     const $elt = $(element);
-    const user_group_id = Number.parseInt($elt.attr("data-user-group-id"), 10);
-    const group = user_groups.get_user_group_from_id(user_group_id);
+    const user_group_id = $elt.attr("data-user-group-id");
```
This might be better called `user_group_id_str` if it's still a string format value.

timabbott commented:
Merged, thanks @sahil839!
